File: benchmarks/cpp/dynamic allocation and free/malloc_free.cpp
#include <iostream>
#include <chrono>
#include <cstdlib>

int main() {
    const size_t N = 1000000;

    volatile size_t counter = 0; // prevents extra optimization

    auto start = std::chrono::high_resolution_clock::now();

    for (size_t i = 0; i < N; i++) {
        int* p = (int*)malloc(sizeof(int));
        counter += (p != nullptr);
        free(p);
    }

    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> t = end - start;

    std::cout << "C++ malloc/free time: " << t.count() << " s\n";
    std::cout << "Counter: " << counter << "\n";

    return 0;
}

File: benchmarks/rust/dynamic allocation and free/malloc_free.rs
use std::time::Instant;
use std::hint::black_box;

fn main() {
    let n: usize = 1000000;
    let mut sink: usize = 0;

    let start = Instant::now();

    for _ in 0..n {
        let p = black_box(Box::new(0u64)); //  prevents elimination
        sink ^= (&*p as *const u64 as usize) & 1;
        drop(p);
    }

    black_box(sink);

    let elapsed = start.elapsed();
    println!(
        "Rust malloc/free time: {:.6} s",
        elapsed.as_secs_f64()
    );
}
